File: TimeCardClientIOS/TimeCardClientIOS/UI/CalendarDetailViewModel.swift
//
//  CalendarDetailViewModel.swift
//  TimeCardClientIOS
//
//  Created by uhimania on 2026/03/13.
//

import Foundation

@Observable
class CalendarDetailViewModel {
    @Observable
    class BreakTime: Identifiable, Equatable {
        var id: UUID
        var start: Date
        var end: Date
        
        init(id: UUID = UUID(), start: Date, end: Date) {
            self.id = id
            self.start = start
            self.end = end
        }
        
        static func == (lhs: CalendarDetailViewModel.BreakTime, rhs: CalendarDetailViewModel.BreakTime) -> Bool {
            lhs.id == rhs.id &&
            lhs.start == rhs.start &&
            lhs.end == rhs.end
        }
    }
    
    @Observable
    class TimeRecord: Identifiable, Equatable {
        var id: UUID
        var checkIn: Date
        var checkOut: Date
        var breakTimes: [BreakTime]
        
        init(id: UUID = UUID(), checkIn: Date, checkOut: Date, breakTimes: [BreakTime] = []) {
            self.id = id
            self.checkIn = checkIn
            self.checkOut = checkOut
            self.breakTimes = breakTimes
        }
        
        func addItem() {
            let breakTime = BreakTime(start: checkIn, end: checkIn)
            breakTimes.append(breakTime)
        }
        
        func deleteItems(indexes: IndexSet) {
            breakTimes.remove(atOffsets: indexes)
        }
        
        func isValid(_ breakTime: BreakTime) -> Bool {
            if breakTime.start > breakTime.end { return false }
            if breakTime.start < checkIn { return false }
            if breakTime.end > checkOut { return false }
            
            let sorted = breakTimes.sorted { $0.start < $1.start }
            for (a, b) in zip(sorted, sorted.dropFirst()) {
                if a.end > b.start {
                    if a.id == breakTime.id || b.id == breakTime.id {
                        return false
                    }
                }
            }
            
            return true
        }
        
        static func == (lhs: CalendarDetailViewModel.TimeRecord, rhs: CalendarDetailViewModel.TimeRecord) -> Bool {
            lhs.id == rhs.id &&
            lhs.checkIn == rhs.checkIn &&
            lhs.checkOut == rhs.checkOut &&
            lhs.breakTimes == rhs.breakTimes
        }
    }
    
    @ObservationIgnored private let repository: CalendarRecordRepository
    
    var date: Date
    var records: [TimeRecord] = []
    var message: String = ""
    private var recordsOriginal: [TimeRecord] = []
    
    init(_ repository: CalendarRecordRepository, _ date: Date) {
        self.repository = repository
        self.date = date
        
        if let record = try? repository.getRecord(year: date.year, month: date.month, day: date.day) {
            self.records = record.records.map { $0.asViewModel() }
            self.recordsOriginal = self.records.copy()
        }
    }
    
    func updateRecord() {
        Task {
            do {
                let record = CalendarRecord(date: date, records: records.map { $0.asRecord() })
                try await repository.updateRecord(record)
                
                self.recordsOriginal = self.records.copy()
            } catch {
                await MainActor.run {
                    self.message = "更新に失敗しました"
                }
            }
        }
    }
    
    func addItem() {
        let record = TimeRecord(checkIn: date, checkOut: date)
        records.append(record)
    }
    
    func deleteItems(indexes: IndexSet) {
        records.remove(atOffsets: indexes)
    }
    
    func isValid() -> Bool {
        if records.contains(where: { !isValid($0) }) { return false }
        
        for rec in records {
            if rec.breakTimes.contains(where: { !rec.isValid($0) }) {
                return false
            }
        }
        
        return records != recordsOriginal
    }
    
    func isValid(_ record: TimeRecord) -> Bool {
        if record.checkIn > record.checkOut { return false }
        if record.checkIn < date { return false }
        
        let sorted = records.sorted { $0.checkIn < $1.checkIn }
        for (a, b) in zip(sorted, sorted.dropFirst()) {
            if a.checkOut > b.checkIn {
                if a.id == record.id || b.id == record.id {
                    return false
                }
            }
        }
        
        return true
    }
}

extension TimeRecord {
    func asViewModel() -> CalendarDetailViewModel.TimeRecord {
        CalendarDetailViewModel.TimeRecord(
            id: self.id,
            checkIn: self.checkIn ?? .now,
            checkOut: self.checkOut ?? .now,
            breakTimes: self.breakTimes.map { $0.asViewModel() }
        )
    }
}

extension TimeRecord.BreakTime {
    func asViewModel() -> CalendarDetailViewModel.BreakTime {
        CalendarDetailViewModel.BreakTime(
            id: self.id,
            start: self.start ?? .now,
            end: self.end ?? .now
        )
    }
}

extension CalendarDetailViewModel.TimeRecord {
    func asRecord() -> TimeRecord {
        TimeRecord(
            id: self.id,
            checkIn: self.checkIn,
            checkOut: self.checkOut,
            breakTimes: self.breakTimes.map { $0.asBreakTime() }
        )
    }
}

extension CalendarDetailViewModel.BreakTime {
    func asBreakTime() -> TimeRecord.BreakTime {
        TimeRecord.BreakTime(
            id: self.id,
            start: self.start,
            end: self.end
        )
    }
}

extension [CalendarDetailViewModel.TimeRecord] {
    func copy() -> [CalendarDetailViewModel.TimeRecord] {
        self.map { $0.copy() }
    }
}

extension CalendarDetailViewModel.TimeRecord {
    func copy() -> CalendarDetailViewModel.TimeRecord {
        CalendarDetailViewModel.TimeRecord(
            id: self.id,
            checkIn: self.checkIn,
            checkOut: self.checkOut,
            breakTimes: self.breakTimes.map { $0.copy() }
        )
    }
}

extension CalendarDetailViewModel.BreakTime {
    func copy() -> CalendarDetailViewModel.BreakTime {
        CalendarDetailViewModel.BreakTime(
            id: self.id,
            start: self.start,
            end: self.end
        )
    }
}

File: TimeCardClientAndroid/app/src/main/java/jp/uhimania/timecardclientandroid/ui/CalendarDetailViewModel.kt
package jp.uhimania.timecardclientandroid.ui

import androidx.annotation.StringRes
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProvider.AndroidViewModelFactory.Companion.APPLICATION_KEY
import androidx.lifecycle.createSavedStateHandle
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import jp.uhimania.timecardclientandroid.R
import jp.uhimania.timecardclientandroid.TimeCardClientApplication
import jp.uhimania.timecardclientandroid.data.BreakTime
import jp.uhimania.timecardclientandroid.data.CalendarRecord
import jp.uhimania.timecardclientandroid.data.CalendarRecordRepository
import jp.uhimania.timecardclientandroid.data.DefaultCalendarRecordRepository
import jp.uhimania.timecardclientandroid.data.TimeRecord
import jp.uhimania.timecardclientandroid.data.day
import jp.uhimania.timecardclientandroid.data.month
import jp.uhimania.timecardclientandroid.data.year
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.util.Date
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

data class CalendarDetailUiState(
    val date: Date = Date(),
    val records: List<TimeRecord> = listOf(),
    val valid: Boolean = true,
    val isLoading: Boolean = false,
    @param:StringRes val message: Int? = null
) {
    data class BreakTime @OptIn(ExperimentalUuidApi::class) constructor(
        val id: String = Uuid.random().toString(),
        val start: Date = Date(),
        val end: Date = Date(),
        val valid: Boolean = true
    )

    data class TimeRecord @OptIn(ExperimentalUuidApi::class) constructor(
        val id: String = Uuid.random().toString(),
        val checkIn: Date = Date(),
        val checkOut: Date = Date(),
        val breakTimes: List<BreakTime> = listOf(),
        val valid: Boolean = true
    )
}

sealed class CalendarDetailUiEvent {
    data object AddTimeRecord : CalendarDetailUiEvent()
    data class UpdateTimeRecord(val id: String, val checkIn: Date, val checkOut: Date) : CalendarDetailUiEvent()
    data class RemoveTimeRecord(val id: String) : CalendarDetailUiEvent()
    data class AddBreakTime(val timeRecordId: String) : CalendarDetailUiEvent()
    data class UpdateBreakTime(val timeRecordId: String, val breakTimeId: String, val start: Date, val end: Date) : CalendarDetailUiEvent()
    data class RemoveBreakTime(val timeRecordId: String, val breakTimeId: String) : CalendarDetailUiEvent()
    data object SaveChanges : CalendarDetailUiEvent()
    data object MessageShown : CalendarDetailUiEvent()
}

class CalendarDetailViewModel(
    private val calendarRecordRepository: CalendarRecordRepository,
    savedStateHandle: SavedStateHandle
) : ViewModel() {
    private val date = Date(savedStateHandle[NavigationArgs.DATE_ARG] ?: 0)

    private val _uiState = MutableStateFlow(CalendarDetailUiState(date = date))
    val uiState = _uiState.asStateFlow()

    init {
        _uiState.update { it.copy(isLoading = true) }
        viewModelScope.launch {
            val record = calendarRecordRepository.getRecord(date.year(), date.month(), date.day())
            _uiState.update {
                validate(record.date, record.records.map { it.asDetailUiState() })
                    .copy(isLoading = false)
            }
        }
    }

    fun onEvent(event: CalendarDetailUiEvent) {
        when (event) {
            is CalendarDetailUiEvent.AddTimeRecord -> addTimeRecord()
            is CalendarDetailUiEvent.UpdateTimeRecord -> updateTimeRecord(event.id, event.checkIn, event.checkOut)
            is CalendarDetailUiEvent.RemoveTimeRecord -> removeTimeRecord(event.id)
            is CalendarDetailUiEvent.AddBreakTime -> addBreakTimeTo(event.timeRecordId)
            is CalendarDetailUiEvent.UpdateBreakTime -> updateBreakTime(event.timeRecordId, event.breakTimeId, event.start, event.end)
            is CalendarDetailUiEvent.RemoveBreakTime -> removeBreakTime(event.timeRecordId, event.breakTimeId)
            is CalendarDetailUiEvent.SaveChanges -> saveChanges()
            is CalendarDetailUiEvent.MessageShown -> messageShown()
        }
    }

    private fun addTimeRecord() {
        val list = _uiState.value.records.toMutableList()
        val date = _uiState.value.date
        list.add(CalendarDetailUiState.TimeRecord(checkIn = date, checkOut = date))
        _uiState.update { validate(_uiState.value.date, list) }
    }

    private fun updateTimeRecord(id: String, checkIn: Date, checkOut: Date) {
        val list = _uiState.value.records.toMutableList()
        val index = list.indexOfFirst { it.id == id }
        if (index != -1) {
            list[index] = list[index].copy(checkIn = checkIn, checkOut = checkOut)
            _uiState.update { validate(_uiState.value.date, list) }
        }
    }

    private fun removeTimeRecord(id: String) {
        val list = _uiState.value.records.toMutableList()
        list.removeAll { it.id == id }
        _uiState.update { validate(_uiState.value.date, list) }
    }

    private fun addBreakTimeTo(timeRecordId: String) {
        val timeList = _uiState.value.records.toMutableList()
        val timeIndex = timeList.indexOfFirst { it.id == timeRecordId }
        if (timeIndex != -1) {
            val breakList = timeList[timeIndex].breakTimes.toMutableList()
            val date = timeList[timeIndex].checkIn
            breakList.add(CalendarDetailUiState.BreakTime(start = date, end = date))
            timeList[timeIndex] = timeList[timeIndex].copy(breakTimes = breakList)
            _uiState.update { validate(_uiState.value.date, timeList) }
        }
    }

    private fun updateBreakTime(timeRecordId: String, breakTimeId: String, start: Date, end: Date) {
        val timeList = _uiState.value.records.toMutableList()
        val timeIndex = timeList.indexOfFirst { it.id == timeRecordId }
        if (timeIndex != -1) {
            val breakList = timeList[timeIndex].breakTimes.toMutableList()
            val breakIndex = breakList.indexOfFirst { it.id == breakTimeId }
            if (breakIndex != -1) {
                breakList[breakIndex] = breakList[breakIndex].copy(start = start, end = end)
                timeList[timeIndex] = timeList[timeIndex].copy(breakTimes = breakList)
                _uiState.update { validate(_uiState.value.date, timeList) }
            }
        }
    }

    private fun removeBreakTime(timeRecordId: String, breakTimeId: String) {
        val timeList = _uiState.value.records.toMutableList()
        val timeIndex = timeList.indexOfFirst { it.id == timeRecordId }
        if (timeIndex != -1) {
            val breakList = timeList[timeIndex].breakTimes.toMutableList()
            breakList.removeAll { it.id == breakTimeId }
            timeList[timeIndex] = timeList[timeIndex].copy(breakTimes = breakList)
            _uiState.update { validate(_uiState.value.date, timeList) }
        }
    }

    private fun validate(date: Date, records: List<CalendarDetailUiState.TimeRecord>): CalendarDetailUiState {
        val invalidTimeRecordIds = mutableSetOf<String>()
        val timeList = records.sortedBy { it.checkIn }
        timeList.filter { !isValid(date, it) }
            .forEach { invalidTimeRecordIds.add(it.id) }

        for ((a, b) in timeList.zip(timeList.drop(1))) {
            if (a.checkOut > b.checkIn) {
                invalidTimeRecordIds.add(a.id)
                invalidTimeRecordIds.add(b.id)
            }
        }

        val invalidBreakTimeIds = mutableSetOf<String>()
        for (rec in timeList) {
            val breakList = rec.breakTimes.sortedBy { it.start }
            breakList.filter { !isValid(rec, it) }
                .forEach { invalidBreakTimeIds.add(it.id) }

            for ((a, b) in breakList.zip(breakList.drop(1))) {
                if (a.end > b.start) {
                    invalidBreakTimeIds.add(a.id)
                    invalidBreakTimeIds.add(b.id)
                }
            }
        }

        val list = records.map { rec ->
            val breakList = rec.breakTimes.map { it.copy(valid = !invalidBreakTimeIds.contains(it.id)) }
            rec.copy(
                breakTimes = breakList,
                valid = !invalidTimeRecordIds.contains(rec.id)
            )
        }

        return _uiState.value.copy(
            date = date,
            records = list,
            valid = invalidTimeRecordIds.isEmpty() && invalidBreakTimeIds.isEmpty()
        )
    }

    private fun isValid(date: Date, record: CalendarDetailUiState.TimeRecord): Boolean {
        if (record.checkIn > record.checkOut) return false
        if (record.checkIn < date) return false

        return true
    }

    private fun isValid(timeRecord: CalendarDetailUiState.TimeRecord, breakTime: CalendarDetailUiState.BreakTime): Boolean {
        if (breakTime.start > breakTime.end) return false
        if (breakTime.start < timeRecord.checkIn) return false
        if (breakTime.end > timeRecord.checkOut) return false

        return true
    }

    private fun saveChanges() {
        viewModelScope.launch {
            try {
                val record = CalendarRecord(
                    date = _uiState.value.date,
                    records = _uiState.value.records.map { it.asTimeRecord() }
                )
                calendarRecordRepository.updateRecord(record)
            } catch (_: Exception) {
                _uiState.update { it.copy(message = R.string.error_update_record_failed) }
            }
        }
    }

    private fun messageShown() {
        _uiState.update { it.copy(message = null) }
    }

    companion object {
        val Factory: ViewModelProvider.Factory = viewModelFactory {
            initializer {
                val app = this[APPLICATION_KEY] as TimeCardClientApplication
                val repository = DefaultCalendarRecordRepository(app.networkDataSource, app.localDataSource)
                val handle = createSavedStateHandle()
                CalendarDetailViewModel(repository, handle)
            }
        }
    }
}

fun TimeRecord.asDetailUiState(): CalendarDetailUiState.TimeRecord {
    return CalendarDetailUiState.TimeRecord(
        id = this.id,
        checkIn = this.checkIn ?: Date(),
        checkOut = this.checkOut ?: Date(),
        breakTimes = this.breakTimes.map { it.asDetailUiState() }
    )
}

fun BreakTime.asDetailUiState(): CalendarDetailUiState.BreakTime {
    return CalendarDetailUiState.BreakTime(
        id = this.id,
        start = this.start ?: Date(),
        end = this.end ?: Date()
    )
}

fun CalendarDetailUiState.TimeRecord.asTimeRecord(): TimeRecord {
    return TimeRecord(
        id = this.id,
        checkIn = this.checkIn,
        checkOut = this.checkOut,
        breakTimes = this.breakTimes.map { it.asBreakTime() }
    )
}

fun CalendarDetailUiState.BreakTime.asBreakTime(): BreakTime {
    return BreakTime(
        id = this.id,
        start = this.start,
        end = this.end
    )
}
